### ISSUE TITLE
Add Mini-Map to Skin

### ISSUE THREAD

**Issue by mkoch227** [mkoch227/ets2-mobile-route-advisor#2]
I tried using the Google Maps API by obtaining lat/lon pairs in relationship to X/Z pairs; however it didn't work too well, most likely since google maps requires a square map, and funbit's is not square.

Possible options:
 - Use funbit's image as a regular JPG and try to line up lat/lon with pixels.
 - Try to reverse-engineer (or just ask for source code) of ets2map.com's map and modify it for single-player
 - ....something else??

**Comment by denilsonsa** [comment]
FYI: Funbit posted [his map image and the code to convert in-game coordinates to pixels](http://forum.scssoft.com/viewtopic.php?f=41&t=186779).

Issue: the image is too large 19200x18688 (almost 359 megapixels) and can't be loaded directly into the browser. It must be sliced into smaller portions and loaded on-demand.

Yeah, Google Maps API may work, but it may be overkill, it is possible to write code to load those sliced images directly, without any external library/service dependency.

**Comment by mkoch227** [comment]
Yep I was having the same issue with Funbit's newer map. If I was lucky, the map would finally render after about 10 seconds of waiting. Trying to nudge the position would take another 10 seconds. However, most of the time my browser tab would simply crash.

I was thinking of splitting the images up as well; however I just haven't had the time to actually dedicate a good chunk of time on it :grinning: . Ideally the splitting would be dynamic, creating a new slice every time the position changes.

**Comment by denilsonsa** [comment]
Short version: I've managed to easily slice the image into small tiles. All we need is the JavaScript code to render these tiles on screen, and maybe to pan and zoom. Do you know any self-contained library for this?

---

Long version:

You can use [vips][1] to slice it:

    vips dzsave \
        map-medium-dark-final.png \
        output_directory \
        --layout google \
        --suffix .png[compression=9] \
        --background '0 0 0 0' \
        --vips-progress

You can change `--layout` to either `dz`, `zoomify` or `google`. By default it has `--tile-size 256`, which seems reasonable. The `--background` defines how to pad the tiles that are smaller than the tile size (i.e., the right-most and bottom-most tiles when zoomed out), receiving red/green/blue/alpha values as numbers from 0 to 255, separated by tab, space or semicolon.

To further reduce the generated PNG size, you may want to use [`zopflipng_in_place`][2] script I've written, which calls [zopfli][3] binary.

If you don't have any Linux (or Mac?) machine nearby, just send me an e-mail and I can send you the sliced images.


Links I've read (saved here for future reference, if needed):
* https://stackoverflow.com/questions/10542246/imagemagick-crop-huge-image
* http://libvips.blogspot.com.br/2013/03/making-deepzoom-zoomify-and-google-maps.html
* http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/libvips/VipsForeignSave.html#vips-dzsave
* https://code.google.com/p/virtualmicroscope/wiki/SlideTiler


[1]: https://github.com/jcupitt/libvips
[2]: https://bitbucket.org/denilsonsa/small_scripts/src/default/zopflipng_in_place
[3]: https://github.com/google/zopfli

**Comment by denilsonsa** [comment]
I did a demo using Leaflet: https://gist.github.com/denilsonsa/d0cd6c2c678cb83199df

Leaflet does not support [map rotation](https://leaflet.uservoice.com/forums/150880-ideas-and-suggestions-for-leaflet/suggestions/5587738-map-rotation), nor [fluid zoom](https://leaflet.uservoice.com/forums/150880-ideas-and-suggestions-for-leaflet/suggestions/3312314-continuous-zoom).

It seems OpenLayers support rotation.

**Comment by mkoch227** [comment]
Thanks for looking into this a bit :smile: . I'll try to take a look at both of your comments in a bit more detail sometime tomorrow evening and see how it all looks.  It sounds like Leaflet will be the better choice for now since it's **much** smaller than OpenLayers.

Looking at the documentation though for leaflet, it looks like the pin's location is defined by latitude/longitude coordinates.  Out of curiosity, how did you handle converting from lat/lon to pixels?

**Comment by denilsonsa** [comment]
I have a split opinion regarding Leaflet and OpenLayers. I like how small and simple Leaflet is, but I really like the idea of the map rotating according to the in-game truck orientation (although without any 3D perspective).

I envision the mobile route advisor map could operate in 3 modes:

* free-view (click/touch and drag to pan the map)
* follow truck fixed rotation (center on the truck, fix the rotation so North is up)
* follow truck including orientation (center on the truck, up is the direction the truck is pointing).

In any of these modes, the user could use the zoom buttons as desired.

Regarding the coordinates… The image we are using has an arbitrary perpective and an arbitrary real-world position. However, we don't care about real-world here. Both mapping libraries support some way to define a custom projection and use custom images, so I defined one that maps lat/lon directly to the image pixels. I bumped into some bugs during the conversion, and I'm not sure the code is 100% correct. Also, on both libraries usually the coordinates increase upwards, which is opposite of the image pixels, so some conversion is needed.

Maybe this "forced" conversion I wrote is not appropriate. In fact, it is buggy in Leaflet (maybe I should have used `L.CRS.Simple` instead of writing my own). I'm not really familiar with those libraries, I may have messed up something.

**Comment by mkoch227** [comment]
Well I got a chance to look at the proof-of-concepts you provided me, and I like how it works. I tried adding the OpenLayers version into the skin; however it's becoming quite difficult trying to add all of the Javascript into a single `.js` file (I can try importing additional JS files; however it becomes quite difficult (see Funbit/ets2-telemetry-server#64).  I'll see if I can get this imported properly soon. I know many people are eager to see it implemented :grinning:

**Comment by mkoch227** [comment]
Well.... I got *something* started. Needless to say it looks terrible and needs improving.

![2015-08-23 17_07_54-start](https://cloud.githubusercontent.com/assets/333430/9430195/f03aa068-49b9-11e5-8ce0-d6066f474bbd.png)

**Comment by mkoch227** [comment]
Ok so I got the map to update the marker based on the current user's position; however the dropped pin doesn't seem to be in the right spot. I'll keep playing around with the code, but if you'd like, you can check out the current state on the **`add-mini-map`** branch.

**Comment by denilsonsa** [comment]
I believe these are the next steps:

* [ ] Merge pull request #19 :wink:
* [ ] Add a button to OpenLayers to toggle "always center on truck".
* [ ] Add a button (it can be the same one) to toggle rotating the map based on the in-game orientation.
* [ ] Rewrite/redesign the layout to be responsive. I believe this is relatively easy, just time-consuming. I'd suggest ditching the bootstrap and starting from scratch. Well, that's how I'd do, and it would end up with a smaller and cleaner HTML. I'm crazy, anyway. :eyes:
* [ ] Ask Funbit to implement a skin config to disable the auto-sizing. By now, we should already have release-candidate version.

**Comment by mkoch227** [comment]
Thanks for the pull request! I vaguely remember the Y and Z coordinates being flipped (in my opinion, Z should be height, not Y :stuck_out_tongue: ), so that makes sense why my initial version was a bit off.  Hopefully adding two buttons shouldn't be too hard (one for always centered, the other for rotating the map).

In terms of the image, I believe I still have a version of ETS2's actual arrow in the `/img` directory; I just never had a chance to implement it :grinning: 

I'd personally be ok with dropping bootstrap for pure CSS, or at least stripping down bootstrap to just what's needed (the progress bars and such   are nice to have, but if they can be replaced with something similar, I'll be ok with that).  Personally I would like to somehow separate the markup of each panel into separate HTML files and have them added at runtime through JS (to make the pages a bit cleaner).

I'm honestly not sure how much time I'll have to work on this as I resumed college courses this week. Hopefully I'll have some time to work on it :grinning:

**Comment by denilsonsa** [comment]
I've rewritten the layout in commit 4183913a24f0ff397c9011ffbb37e3d6427e0fab.

Check the [most recents commits from my fork](https://github.com/denilsonsa/ets2-mobile-route-advisor/commits/add-minimap). They require this change to the server: https://github.com/Funbit/ets2-telemetry-server/pull/66.

And if you are in a hurry, [look at the results in this short video](https://youtu.be/26BqmAyqw5g).

**Comment by mkoch227** [comment]
4183913 looks great :grinning:.  I also really like the new icons and how the progress bar is no longer needed.  Feel free to submit a PR whenever you'd like.  Also, if you want, I can add you as a collaborator to the project so you can push directly to the main repo, rather than pushing to a fork and submitting PRs. Just let me know :smile:

**Comment by denilsonsa** [comment]
For future reference, I've saved a collection of ETS2-related scripts and stuff at <https://github.com/denilsonsa/ets2-stuff>. Some of that stuff might be useful whenever someone needs to re-create the map tiles. (But mostly it is just a repo where I can dump stuff.)

Also, you don't need to wait for a pull request, you can grab my changes (or anyone's changes) anytime:

    git remote add denilsonsa https://github.com/denilsonsa/ets2-mobile-route-advisor.git
    git fetch denilsonsa
    git checkout add-minimap  # Checking out your own branch.
    git merge denilsonsa/add-minimap  # Merging with my branch.
    git push

Not sure if you knew that. :)

**Comment by denilsonsa** [comment]
I believe everything I wanted to have in the map [has been implemented](https://github.com/mkoch227/ets2-mobile-route-advisor/pull/20). I believe this can be an alpha/beta version, and very soon a release candidate.

Known issues/next steps:

* [ ] Update `README.md`.
* [ ] Update screenshots.
* [ ] Wait until https://github.com/Funbit/ets2-telemetry-server/pull/66 gets merged.
* [ ] Test on Safari and iOS.
* [ ] `navigation-estimatedTime` is shown as `NaN:NaN` in Firefox. This also happens in `master` branch.

Low-priority stuff (that will not be changed anytime soon):

* There is no "animation" in the map. Clicking on the rotation button changes the view instantly. This is not a good UX, but it is also a limitation and side-effect of <https://stackoverflow.com/q/32868671/>.
* The map image could have been higher resolution, but that's not an easy task. See also: <https://github.com/denilsonsa/ets2-stuff>.

Some strings are not translated (I believe translations are low priority):

* [ ] Week days (should they be abbreviated? like in-game?)
* [ ] `job-remainingTime` (but, oddly enough, `game-nextRestStopTime` is translated)
* [ ] "Expected" text (that I introduced, copying the same text from the game)
* [ ] "About" page (should it be translated? Or is it too much work for little gain?

I'd really like if someone else (maybe you!) tackled these issues, specially the first three, which are blockers to an actual release. All the others are just "polish": they are nice to have, but not essential.

**Comment by mkoch227** [comment]
I just got a chance to look at and test out your pull request, and the changes look awesome :grinning: .  I'll merge them in soon.

In response to all of your checkbox points:
 - I'll update the `README` and screenshots within the next week or so, and push them to a `3-0-0` branch.
 - Hopefully Funbit/ets2-telemetry-server#66 gets merged soon. If not, worst case scenario is that we can build our own version of the telemetry server and have users download the custom version instead.
 - Unfortunately I do not own any iOS devices. We can either go ahead as-is and hope for the best, or try and find someone with an iOS device who can test out the skin.
 - I just created #21 in reference to the `NaN:NaN` issue in Firefox.

 - I'm not too concerned about the lack of a fluid transition between auto-rotating and keeping a fixed position.  In the actual game, the user is only given a few pre-determined zoom levels, and once you zoom out to a certain point, the map switches to the fixed position unless you zoom in again.
 - In terms of screen resolution, I think the current resolution is fine to begin with. If we can somehow get a larger size, we can do so at a later point.

 - I'm ok with abbreviating the day of the week, and we can also add them to the translation `.json`s as well.  Another option is to look into JavaScript's `Date.prototype.toLocaleString()`, which I believe uses the user's browser locale to build the date. It also looks like it supports several options on how to parse the date.
 - There is a `TimeRemaining` language string, so I'll have to double-check that it's actually being used.
 - We can easily add a new language string for "Expected", and we can remove "Deadline" if it is no longer being used.
 - I personally think that translating the about tab would be too much work. I have a feeling hardly anyone looks at that tab, and it would take a lot of screen space in the language `.json`s.

**Comment by denilsonsa** [comment]
I believe the "Expected" line in-game shows both the minimum time and the maximum time for arrival (look at [the guide for this achievement](http://astats.astats.nl/astats/Steam_Achievement_Info.php?AchievementID=15&AppID=227300#tabs-2)). I think we have access to the maximum time (i.e. the deadline), but not the minimum one. If we don't, that's something we may ask the developers to add to the API.

One last thing I noticed: although the gradient colors currently in use look good, they don't provide enough contrast with the white text. I noticed this issue while playing the game trying to quickly see the speed in my mobile phone. We can change the gradients to look exactly like in-game. (#1e1e1e-#505050; #181818-#404040; the bottom one is #2c2c2c-#535353; look at the button gradient for an example)

For help with testing on Apple devices and with translations, we may try asking in [/r/trucksim](https://www.reddit.com/r/trucksim) and in [/r/EuroTruck2](https://www.reddit.com/r/EuroTruck2).

In a far away future, we may have a list of in-game coordinates and places, so that we can add a map layer full of markers.

**Comment by denilsonsa** [comment]
Regarding providing an alternate telemetry server download, it is trivial: https://github.com/denilsonsa/ets2-telemetry-server/archive/no-auto-size.zip

**Comment by mkoch227** [comment]
@denilsonsa it looks Safari does have some issues (both on OS X and through the iOS simulator). See #26

**Comment by denilsonsa** [comment]
I've seen the latest commits, great work! :)

**Comment by raflix261** [comment]
I have a problem with version 3.0.0 not see the preview icon .
![screenshot_1](https://cloud.githubusercontent.com/assets/8672101/10267804/45eccd00-6aa4-11e5-8f3e-c2c24a7f88bf.png)

And when I run error pops up. I use Chrome.
![screenshot_2](https://cloud.githubusercontent.com/assets/8672101/10267807/85c3bd62-6aa4-11e5-9dcd-480c81876fa8.png)
The same effect is on Firefox and do not know what to do.

**Comment by denilsonsa** [comment]
@raflix261 , what system?

Is this related? https://github.com/Funbit/ets2-telemetry-server/pull/63

Also, our version 3.0.0 REQUIRES https://github.com/Funbit/ets2-telemetry-server/pull/66 , which can be downloaded from https://github.com/denilsonsa/ets2-telemetry-server/archive/no-auto-size.zip

**Comment by raflix261** [comment]
@denilsonsa The system is Windows 10. With the trial version which lent the map I had no problem. The same is true for the second version of the server you specified.

**Comment by mkoch227** [comment]
> In a far away future, we may have a list of in-game coordinates and places, so that we can add a map layer full of markers.

I did some peeking around at kat_pw's ets2map.com, and I found this handy JSON that is used to output POIs on the map (such as fuel, hotel, and even speed cameras). I copied the JSON and tossed it into the gist below.  I don't know if there are any limitations as to us using it; however I thought I'd share it since it looks like it also uses in-game coordinates.  However, I don't see this being a necessary feature for 3.0.0.

https://gist.github.com/mkoch227/bdadae22eb616b4b9dfb

**Comment by mkoch227** [comment]
> I believe the "Expected" line in-game shows both the minimum time and the maximum time for arrival (look at the guide for this achievement). I think we have access to the maximum time (i.e. the deadline), but not the minimum one. If we don't, that's something we may ask the developers to add to the API.

Correct, only the deadline time is provided to us.  From my gameplay, I haven't noticed any adverse effects of arriving before the dropoff window (I've never been penalized).  I also got the achievement in question, but I cannot say if I was within the time window or not.  Anyhoo, I find that to be a low priority issue in my opinion.

**Comment by mkoch227** [comment]
@denilsonsa is there anything else that needs to be done before the initial release? (Besides getting https://github.com/Funbit/ets2-telemetry-server/pull/66 merged in)

**Comment by denilsonsa** [comment]
@raflix261 commented on issue #38 regarding an untranslated time, but I haven't checked if it is already fixed.

Regarding the code, I can't think of anything else.

Sure, maybe we can improve the translations (to make them look exactly like in-game), but this can be done after 3.0.0.

---

Regarding README (the front page of the the GitHub repository), instructions on how to download and install the skin. One of the steps should be to make sure the directory has the right name.

Of course, also mentioning the https://github.com/Funbit/ets2-telemetry-server/pull/66 (and where to download the alternate version).

---

Regarding the website, essentially the same stuff as the README.

---

Footnote: I haven't tested the latest commits yet, but they should be okay.

**Comment by mkoch227** [comment]
Thanks. I've created two new issues (#42 and #43) for the updates you recommended. Since the map is pretty much done, I'll close this issue and make sure everything for the minimap is in the `3-0-0` branch.

**Comment by mkoch227** [comment]
Everything in `add-minimap` is in `3-0-0`. Closing this issue and deleting the `add-minimap` branch.
